### PR TITLE
Merge compounds

### LINKED ISSUE #14: Merge Individuals

### LINKED ISSUE #15: Merge Formulations

### LINKED ISSUE #16: Merge compounds

### LINKED ISSUE #17: Merge Protocols

### LINKED ISSUE #18: Merge Simulations

### LINKED ISSUE #24: Merge Expression profiles

### LINKED ISSUE #28: Merge Populations

### LINKED ISSUE #30: Handle duplicated Names

### LINKED ISSUE #34: Harmonize Snapshot Versions Before Merging them
JSON structure can be imcompatible between Versions.
So **if versions are not all the same** add an extra step:
Convert user snapshots to project (pksim5) then back to snapshots.

This way, all output snapshots will have same version and then can be merged together.

### LINKED ISSUE #36: Print Compound Parameters

### LINKED ISSUE #37: Print DDI Parameters

### LINKED ISSUE #39: Import DDI Snapshot

### PR DESCRIPTION
**Pull request by Felixmil**
closes #14 
closes #15 
closes #16 
closes #17 
closes #18 
closes #24 
closes #28 
closes #30 
closes #34 
closes #36 
closes #37 
closes #39

### CODE REVIEW

File: R/ddi.R [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+      snapshots <- list(...)
+
+      snapshot_versions <- map(snapshots, ~ .x$data$Version) %>% list_c()
+
+      # If Snapshots have different versions
+      if (length(unique(snapshot_versions)) > 1) {
+        cli_process_start("Multiple versions detected. Converting to the latest version.")
+
+        temp_dir <- tempfile()
+        dir.create(temp_dir)
```
Comment (dlefaudeux) [inline R/ddi.R:70]: Maybe make sure the temp_dir gets cleaned at some point.

File: R/model_library.R [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+#'
+#' @noRd
+#' @keywords internal
+get_osp_model_library <- function() {
+  check_internet()
+
+  # If the data is not already cached in `the` package environment, download it
+  if (is.null(the$ model_library_data)) {
+    cli_process_start(msg = "Listing Compounds Building Blocks available in {style_hyperlink(text = 'OSP repositories', url = model_library_url)}")
+    model_library_data_raw <- jsonlite::fromJSON(model_library_url, simplifyDataFrame = FALSE)$Templates
```
Comment (dlefaudeux) [inline R/model_library.R:14]: maybe also need simplifyVector = FALSE here too

**Review by rengelke — COMMENTED** [review]
Amazing work so far! 

Some minor comments

`Snapshot`:

- `get_names` and `print` methods need tests.
- Methods like `compoundsNames`, `individualsNames`, and `populationsNames` all call `private$get_names`. This could be refactored into a single method that accepts a field name.
- `data` field is populated: `self$data <- private$read_json(self$source)` . Do we need to or can we validate the data? When accessing parts of the `self$data` later, what happens if certain fields don't exist? 

`DDI` Class:
When checking sections to merge for duplicates, the logic keeps the first element for duplicate entries. Is it safe to assume that the first entry is always the correct one?

Not related to this PR:
Why importing all packages in zzz.R? The are a lot of namespace conflicts when loading the package.
My personal preference is to use explicit namespaces when calling external functions. It keeps the namespace clean and makes the code more understandable.

File: R/ddi.R (hunk 1 above)
Comment (Felixmil) [reply R/ddi.R:70]: When session is closed, all tempfile() are deleted

File: R/model_library.R (hunk 2 above)
Comment (Felixmil) [reply R/model_library.R:14]: done

**Comment by rengelke** [comment]
> > * Methods like `compoundsNames`, `individualsNames`, and `populationsNames` all call `private$get_names`. This could be refactored into a single method that accepts a field name.
> 
> `private$get_names` is already a generic function that all other use. So I'm not sure what you mean here.
> 
I see, so this are just wrappers to create user friendly function names.

> > * `data` field is populated: `self$data <- private$read_json(self$source)` . Do we need to or can we validate the data? When accessing parts of the `self$data` later, what happens if certain fields don't exist?
> 
> I would simply return NULL and rest of the logic takes care of NULL values during merge (they are removed)
> 
Understand that if fields are not present in `self$data` it will return NULL
> > When checking sections to merge for duplicates, the logic keeps the first element for duplicate entries. Is it safe to assume that the first entry is always the correct one?
> 
> First element should be the victim compound, which is _a priori_ the most important one. However, a warning is displayed to the user. He just has to rename the item in its PKSIM project of in the json file.
> 
Alright, then this is considered.
> > Why importing all packages in zzz.R? The are a lot of namespace conflicts when loading the package.
> > My personal preference is to use explicit namespaces when calling external functions. It keeps the namespace clean and makes the code more understandable.
> 
> I agree and reduced them to the minimum for most used functions. I did that because writing piped/nested commands with several `purrr::` or `cli::` makes the code really messy in the end.
Yes, one disadvantage is that the code is more verbose.

**Review by rengelke — APPROVED** [review]

**Pull request merged**